### ISSUE TITLE
CI: samples/userspace/hello_world_user/sample.helloworld fails on qemu_xtensa_mmu

### ISSUE THREAD

**Issue by henrikbrixandersen** [zephyrproject-rtos/zephyr#66029]
**Describe the bug**
The sample `samples/userspace/hello_world_user/sample.helloworld` fails in CI on `qemu_xtensa_mmu`. Git bisect points towards commit a36e39c2a65f3705832875053f8a4e13320dcd4c, which is quite old - but does fit the error.

**To Reproduce**
1. `west twister -p qemu_xtensa_mmu -s samples/userspace/hello_world_user/sample.helloworld`
2. See error

**Expected behavior**
CI passes.

**Impact**
CI broken on `main`.

**Logs and console output**
```
*** Booting Zephyr OS build zephyr-v3.5.0-2432-g8f9fd239eb26 ***

Backtrace:0x00101d08:0x0014def0 0x00101cfa:0x0014df60 0x001005ce:0x0014dfb0 0x00100789:0x0014dfd0 0x4010172d:0x0014dff0  |<-CORRUPTED

```

**Environment (please complete the following information):**
 - Commit SHA: 08e42b147edb4d1f52d297b02df8497b0b962c9d

**Comment by andyross** [comment]
Sneak myself in as an assignee to get visibility.

Weird that it's SDK-specific.  Paging behavior isn't generally related to code generation; it's just about semantic behavior: you hit the same pages in the same order either way, though maybe it's purely an issue of code size or something, like one version is a page bigger or smaller and hits an off-by-one somewhere?

**Comment by dcpleung** [comment]
Looks like QEMU is returning incorrect physical address when accessing a variable in the libc partition:

```
get_pte: autorefill(00109020): PTE va = 20000424, pa = 0010c424
get_physical_addr_mmu: autorefill(00109020): 00109000 -> 00109006
xtensa_cpu_tlb_fill(00109020, 1, 0) -> 00109020, ret = 0
xtensa_cpu_tlb_fill(00109028, 1, 0) -> 00109028, ret = 0
xtensa_cpu_tlb_fill(00109014, 0, 2) -> 00103050, ret = 26
```

Note that `0x109014` somehow mapped to `0x103050` which is incorrect as we always identity map memory pages.

**Comment by andyross** [comment]
You think it's a qemu bug?  Does it pass in xt-sim?  I guess my gut looking at that would say it was an uninitialized or garbage page in the table?  Can you dump the relevant pages ahead of time to verify qemu is loading different bytes than memory sees?

(For clarity: xtensa_cpu_tlb_fill() is reporting that a request for that address produced the word after the arrow?  What are the two other arguments, maybe ring and instruction/data or something?)

**Comment by andyross** [comment]
OK, found it: https://github.com/qemu/qemu/blob/master/target/xtensa/helper.c#L262

This is even more confusing, I think?  The "ret = 26" implies that the underlying fetch failed for some reason, and as I read it it will then set up a exception trap.  So that third line is actually fine, and I'm betting the 0x103050 value is just junk on the stack, it's not valid because the function failed.

So the question is why the load from 0x109014 failed.   Which is especially weird because the previous line shows a load from  0x109028, which is **in the same page** and should hit the same TLB line?  So why does it work the first time and fail the second?

I **think** (@jcmvbkbc -- interpretation would be greatly appreciated here) that this has to do with the second argument, mmu_idx.  That's compared against a ring value in the underlying xtensa_get_physical_address() function, so it's probably the ring/access level of the fetch?  So maybe the TLB line is stored at a ring of zero and not 2?  (Or three?  Are we using ring 2 and not 3 for userspace?)

Short version: this is consistent with writing the PTE entry from a ring 0 kernel mode mapping, which will populate the TLB at ring zero, then switching to a userspace context that has the same address mapped at ring 2 without doing a TLB flush.  And indeed doing the flush at swap time would indeed fix that.

 Interestingly this is exactly what @ceolin was pointing out was possible.  :)

**Comment by andyross** [comment]
Actually thinking on that it still doesn't add up.  A TLB auto fill should never be done at anything but ring zero, which still sounds like a qemu bug?  Or something we don't understand anyway.

And needless to say PTE data should never be mapped at anything but ring 0, nor touched from userspace.  So still confusing.

**Comment by andyross** [comment]
OK, this is looking like a genuine qemu bug.  Indeed that "mmu_idx" value is interpreted as a ring when it gets to the get_physical_addr_mmu() function here:

https://github.com/qemu/qemu/blob/master/target/xtensa/mmu_helper.c#L820

But that's just wrong, because TLB refill fetches never use anything but ring zero.  The value being passed is (I think) the priviledge level of the original fetch.

Are you set up to build qemu?  It might be as simple as this, if we're lucky:

```patch
diff --git a/target/xtensa/helper.c b/target/xtensa/helper.c
index dbeb97a953..968c05b32c 100644
--- a/target/xtensa/helper.c
+++ b/target/xtensa/helper.c
@@ -269,7 +269,7 @@ bool xtensa_cpu_tlb_fill(CPUState *cs, vaddr address, int size,
     uint32_t page_size;
     unsigned access;
     int ret = xtensa_get_physical_addr(env, true, address, access_type,
-                                       mmu_idx, &paddr, &page_size, &access);
+                                       0, &paddr, &page_size, &access);

     qemu_log_mask(CPU_LOG_MMU, "%s(%08" VADDR_PRIx
                   ", %d, %d) -> %08x, ret = %d\n",
```

Again, hoping Max wanders by to take a look.  I should have some time tomorrow to try this I hope.

**Comment by andyross** [comment]
I did not, I found what looked like a smoking gun, scribbled a patch, tried it and walked away imagining victory.  I **absolutely** don't expect I understand all of what's going on here. :)  Though that tlb_fill() API (which is a function pointer callback implementing a standard qemu thing, so I don't know what exactly it's supposed to do) doesn't seem to be treating with the instruction pointer at all?   It certainly looks like a trap handler for a missing translation.

**Comment by dcpleung** [comment]
Both `mem_protect` and `userspace` failed on the patched QEMU but the one in SDK worked. So using `mmu_idx` seems to be correct on that line.

**Comment by henrikbrixandersen** [comment]
Thank you for the thorough examination of this issue. Anything we can do to unblock CI on `main` while the investigation continues? Can we exclude the platform from the test or similar?

**Comment by andyross** [comment]
My vote would be to exclude the test, yeah.  Though no objections to merging Daniel's flush fix as a workaround, I just want to make sure we don't forget and ship it, it's sort of a performance disaster if it reaches real hardware.

**Comment by ceolin** [comment]
> My vote would be to exclude the test, yeah. Though no objections to merging Daniel's flush fix as a workaround, I just want to make sure we don't forget and ship it, it's sort of a performance disaster if it reaches real hardware.

I rather have Daniel's fix for now, but tbh one line fix with a huge TODO to remove that seems better than something that looks like  valid option (Kconfig option)

**Comment by dcpleung** [comment]
Either way is fine with me. Just need to pick one.

**Comment by henrikbrixandersen** [comment]
Please note: Commit 031c842ecb768a2a98c33bd0259387e990cfece1 needs to be reverted once this is properly fixed.

**Comment by jcmvbkbc-issues** [comment]
I totally missed this issue, but I've got a notification today and tried to reproduce it. I'd say that I couldn't. That is with the SDK 0.16.4 and zephyr 8f9fd239eb26 twister indeed reports a failure, but if I run the test manually I get the following output from the test:

`*** Booting Zephyr OS build zephyr-v3.5.0-2432-g8f9fd239eb26 ***
Hello World from UserSpace! (qemu_xtensa_mmu)`

which then goes into the idle loop. I don't see any of the mmu-related oddities mentioned in the thread.

**Comment by jcmvbkbc-issues** [comment]
I totally missed this issue, but I've got a notification today and tried to reproduce it. I can see a LoadStorePrivilegeCause exception at PC `0x00101d0b` when it tries to load a word from address `0x00109014` from ring 2 and this attempt finds a TLB entry for the page `0x00109xxx` with ASID 1 which maps to ring 0. This entry was loaded by the TLB autorefill while executing basic block at PC `0x001017db` and it corresponds to a variable `_stdout_hook`.

**Comment by jcmvbkbc-issues** [comment]
A virtual address cannot be safely mapped into multiple address spaces which can simultaneously be present in the RASID SR, and since ASID 1 is always there associated with ring 0 it should not be used to access user pages at the same virtual addresses as the userspace.
I'm not exactly sure how zephyr maps kernel and user code and data, but it seems that it tries to share printing code/data between the kernel and userspace at the same virtual address. This is not going to work reliably with kernel in ring 0 unless TLB is flushed on switch to userspace.
